llvm/lib/Target/EVM/EVMStackifyCodeEmitter.h

```diff
@@ -0,0 +1,91 @@
+//===--- EVMStackifyCodeEmitter.h - Create stackified MIR  ------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file transforms MIR to the 'stackified' MIR.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_LIB_TARGET_EVM_EVMSTACKIFYCODEEMITTER_H
+#define LLVM_LIB_TARGET_EVM_EVMSTACKIFYCODEEMITTER_H
+
+#include "EVMStackLayoutGenerator.h"
+#include "EVMSubtarget.h"
+
+namespace llvm {
+
+class MachineInstr;
+class MCSymbol;
+
+class EVMStackifyCodeEmitter {
+public:
+  EVMStackifyCodeEmitter(const StackLayout &Layout, MachineFunction &MF)
+      : Emitter(MF), Layout(Layout), MF(MF) {}
+
+  /// Stackify instructions, starting from the \p EntryBB.
+  void run(CFG::BasicBlock &EntryBB);
+
+private:
+  class CodeEmitter {
+  public:
+    explicit CodeEmitter(MachineFunction &MF)
+        : MF(MF), TII(MF.getSubtarget<EVMSubtarget>().getInstrInfo()) {}
+    int getStackHeight() const;
+    void init(MachineBasicBlock *MBB, int Height);
+    void emitInst(const MachineInstr *MI);
+    void emitSWAP(unsigned Depth);
+    void emitDUP(unsigned Depth);
+    void emitPOP();
+    void emitConstant(const APInt &Val);
+    void emitSymbol(const MachineInstr *MI, MCSymbol *Symbol);
+    void emitConstant(uint64_t Val);
+    void emitFuncCall(const MachineInstr *MI, const GlobalValue *Func,
+                      int StackAdj, bool WillReturn);
+    void emitRet(const MachineInstr *MI);
+    void emitCondJump(const MachineInstr *MI, MachineBasicBlock *Target);
+    void emitUncondJump(const MachineInstr *MI, MachineBasicBlock *Target);
+    void emitLabelReference(const MachineInstr *Call);
+    void finalize();
```

Comment:
Documentation is welcome for `finalize`. The rest are self-explanatory methods.

Reply:
Documented this function.